cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/spark/SparkRuntimeService.java

```diff
@@ -269,112 +261,174 @@ public void apply() throws Exception {
+        if (className.startsWith("org.apache.spark")
+          || className.startsWith("scala")
+          || classPathUrl.toString().contains("spark-assembly")) {
+          return false;
+        }
+        return hadoopClassExcluder.accept(className, classUrl, classPathUrl);
+      }
+    });
+    appBundler.createBundle(tempLocation, SparkProgramWrapper.class, HBaseTableUtilFactory.getHBaseTableUtilClass());
+    return new File(tempLocation.toURI());
   }
 
   /**
-   * Copies the user submitted program jar and flatten it out by expanding all jars in the "/" and "/lib" to top level.
+   * Copy the program jar to local directory.
    *
-   * @param jobJarLocation {link Location} of the user's job
-   * @param context        {@link BasicSparkContext} context of this job
-   * @return {@link Location} where the program jar was copied
-   * @throws IOException if failed to get the {@link Location#getInputStream()} or {@link Location#getOutputStream()}
+   * @return {@link File} of the copied jar in the given target directory
    */
-  private Location copyProgramJar(Location jobJarLocation, BasicSparkContext context) throws IOException {
-    Id.Program programId = context.getProgram().getId();
-    Location programJarCopy = locationFactory.create(String.format("%s.%s.%s.%s.%s.program.jar",
-                                                                   ProgramType.SPARK.name().toLowerCase(),
-                                                                   programId.getNamespaceId(),
-                                                                   programId.getApplicationId(), programId.getId(),
-                                                                   context.getRunId().getId()));
-
-    ByteStreams.copy(Locations.newInputSupplier(jobJarLocation), Locations.newOutputSupplier(programJarCopy));
-    return programJarCopy;
+  private File copyProgramJar(Location programJar, File targetDir) throws IOException {
+    File tempFile = new File(targetDir, "program.jar");
+
+    LOG.debug("Copy program jar from {} to {}", programJar, tempFile);
+    Files.copy(Locations.newInputSupplier(programJar), tempFile);
+    return tempFile;
   }
 
-  private Runnable createCleanupTask(final Iterable<File> files, final Iterable<Location> locations) {
+  /**
+   * Generates an empty JAR file.
+   *
+   * @return The generated {@link File} in the given target directory
+   */
+  private File generateJobJar(File targetDir) throws IOException {
+    File tempFile = new File(targetDir, "dummy.jar");
+    JarOutputStream output = new JarOutputStream(new FileOutputStream(tempFile));
+    output.close();
+    return tempFile;
+  }
+
+  /**
+   * Serialize {@link CConfiguration} to a file.
+   *
+   * @return The {@link File} of the serialized configuration in the given target directory.
+   */
+  private File saveCConf(CConfiguration cConf, File targetDir) throws IOException {
+    File file = new File(targetDir, "cConf.xml");
```

Comment:
should we use constants for these file name? That way it becomes a lot easier to find out where they are used.